File: js/parser.js

// todo delete this file

var REGEX_STR = {
    PLAYER: "([\\w\\.\\-\| ]+) ",
    CHARACTER: "(?:\\(|\\[) *([\\w\\. ]+) *(?:\\)|\\]) ",
    VERSUS: "(?:Vs|vs)\\.? ",
}

var YogaFlameParser = (function(){
    var self = {};
    var fixer = FIXER;

    var GAME_REGEX_STR = {
        SF5: "SF5|SFV|Beta SFV",
        USF4: "USF4",
        SSF4AE2012: "(?:Arcade Edition|AE)(?: Version)? +2012",
        SSF4AE: "Arcade Edition|AE",
        SF3: "SF3",
        SFxT: "SFxT",
        TTT2: "Tekken Tag Tournament 2"
    }
    var GAME_REGEX = {};
    function buildRegex(){
        var rounds = "(?:X[0-9] )?";
        var resolution = " *(?:1080p|720p)";
        var gameRegexArr = [];
        for (var key in GAME_REGEX_STR) {
            if (GAME_REGEX_STR.hasOwnProperty(key)) {
                GAME_REGEX[key] = new RegExp(GAME_REGEX_STR[key]);
                gameRegexArr.push(TOOL.format("{1}", GAME_REGEX_STR[key]));
            }
        }
        var combinedGame = gameRegexArr.join('|');
        var game = TOOL.format(" *({1}) ", combinedGame);
        var regexStr = (
            rounds +
            REGEX_STR.PLAYER +
            REGEX_STR.CHARACTER +
            REGEX_STR.VERSUS +
            REGEX_STR.PLAYER +
            REGEX_STR.CHARACTER +
            game +
            resolution
        );
        return new RegExp(regexStr);
    }
    var REGEX = buildRegex();
    TOOL.log(REGEX);

    function fixGame(game){
        var res = null;
        for (var key in GAME_REGEX) {
            if (GAME_REGEX.hasOwnProperty(key)) {
                res = GAME_REGEX[key].exec(game);
                if (res){
                    return key;
                }
            }
        }
        return game;
    }

    self.parse = function(dataTuple){
        var timestamp = dataTuple[0];
        var id = dataTuple[1];
        var title = dataTuple[2];
        var res = REGEX.exec(title);
        if (!res){
            TOOL.visuLog(title);
            return null;
        }
        return VideoItem(
            timestamp,
            id,
            title,
            fixGame(res[5]),
            fixer.fixCharacters(res[2], res[4]),
            fixer.fixPlayers(res[1], res[3])
        );
    };

    return self;
})();

var OlympicParser = (function(){
    var self = {};
    var fixer = FIXER;

    var GAME_REGEX_STR = {
        SF5: "Street Fighter (?:5 */? *V|5|V)|SFV",
        SFxT: "Street Fighter X Tekken",
        P4AU: "Persona 4 Arena Ultimax",
        MKX: "Mortal Kombat X",
        SC5: "SoulCalibur 5\/V",
        DOA5: "Dead Or Alive 5 Last Round",
        Smash4: "Super Smash Bros Wii U",
        GGXrd: "Guilty Gear Xrd",
        KI: "Killer Instinct",
    }
    var GAME_REGEX = {};
    var REGEX_1 = null;
    var REGEX_2 = null;
    function buildRegex(){
        var endtag = "(?:Wii|Xbox|PS4|-? ?Gameplay).*";
        var gameRegexArr = [];
        for (var key in GAME_REGEX_STR) {
            if (GAME_REGEX_STR.hasOwnProperty(key)) {
                GAME_REGEX[key] = new RegExp(GAME_REGEX_STR[key]);
                gameRegexArr.push(TOOL.format("{1}", GAME_REGEX_STR[key]));
            }
        }
        var combinedGame = gameRegexArr.join('|');
        var game = TOOL.format(" *({1}) *", combinedGame);
        var regexStr1 = (
            REGEX_STR.PLAYER +
            REGEX_STR.CHARACTER +
            REGEX_STR.VERSUS +
            REGEX_STR.PLAYER +
            REGEX_STR.CHARACTER +
            game +
            endtag
        );
        REGEX_1 = new RegExp(regexStr1);
        var regexStr2 = (
            game + ': ' +
            REGEX_STR.PLAYER +
            REGEX_STR.CHARACTER +
            REGEX_STR.VERSUS +
            REGEX_STR.PLAYER +
            REGEX_STR.CHARACTER +
            endtag
        );
        REGEX_2 = new RegExp(regexStr2);
    }
    buildRegex();
    TOOL.log(REGEX_1);
    TOOL.log(REGEX_2);

    function fixGame(game){
        var res = null;
        for (var key in GAME_REGEX) {
            if (GAME_REGEX.hasOwnProperty(key)) {
                res = GAME_REGEX[key].exec(game);
                if (res){
                    return key;
                }
            }
        }
        return game;
    }

    self.parse = function(dataTuple){
        var timestamp = dataTuple[0];
        var id = dataTuple[1];
        var title = dataTuple[2];

        var res = REGEX_1.exec(title);
        if (res){
            return VideoItem(
                timestamp,
                id,
                title,
                fixGame(res[5]),
                fixer.fixCharacters(res[2], res[4]),
                fixer.fixPlayers(res[1], res[3])
            );
        }
        res = REGEX_2.exec(title);
        if (res){
            return VideoItem(
                timestamp,
                id,
                title,
                fixGame(res[1]),
                fixer.fixCharacters(res[3], res[5]),
                fixer.fixPlayers(res[2], res[4])
            );
        }
        // else
        TOOL.visuLog(title);
        return null;
    };

    return self;
})();

File: scala/parser.scala

package fgc.parser

import scala.util.matching.Regex
import scala.collection.immutable.ListMap

import fgc.model.VideoData
import fgc.model.VideoItem
import fgc.channel.YouTubeChannel
import fgc.logger.Logger

object YouTubeChannelParser {
    val Parsers = List(
        YogaFlameParser,
        OlympicGamingParser
    )
}

trait ChannelParser {
    val channel: YouTubeChannel
    def loadVideos(): List[VideoData] = {
        channel.loadFile.values.flatMap { videoItem =>
            val opt = parseVideo(videoItem)
            if (!opt.isDefined){
                Logger.parsingFailure(channel.fileName, videoItem.title)
            }
            opt
        }.toList
    }

    def fixRegex(rawRegexStr: String): Regex = { ("(?i)" + rawRegexStr).r }

    trait VideoParser {
        val regexStr: String
        def parseSuccess(videoItem: VideoItem, regexMatch: List[String]): VideoData {}
    }
    val parsers: List[VideoParser]
    def parseVideo(videoItem: VideoItem): Option[VideoData] = {
        parsers.flatMap { parser =>
            fixRegex(parser.regexStr).findFirstMatchIn(videoItem.title).map { m =>
                parser.parseSuccess(videoItem, m.subgroups)
            }
        }.headOption
    }

    val rPlayer = "(?:[\\w ]* (?:ft |ft\\.))? *([\\w\\.\\-\\| ]+) *"
    val rCharacter = " *(?:\\(|\\[) *([\\w\\. ]+) *(?:\\)|\\]) *"
    val rVersus = "(?:vs)\\.? "
    val rGameMap: Map[String, String]
    def rGame(): String = {
        val rGames = rGameMap.values.mkString("|")
        s" *($rGames) *"
    }

    def fixGame(rawGame: String): String = {
        rGameMap.flatMap { case (key, gameRegexStr) =>
            fixRegex(gameRegexStr).findFirstIn(rawGame).map { m => key }
        }.head
    }
    def fixCharacters(char1: String, char2: String): List[List[String]] = {
        List(List(char1), List(char2))
    }
    def fixPlayers(player1: String, player2: String): List[String] = {
        List(player1, player2)
    }
}

object YogaFlameParser extends ChannelParser {
    val channel = YouTubeChannel.YogaFlame

    val rGameMap = ListMap(
        "SF5" -> "SF5|SFV|Beta SFV",
        "USF4" -> "USF4",
        "SSF4AE2012" -> "(?:Arcade Edition|AE)(?: Version)? +2012",
        "SSF4AE" -> "Arcade Edition|AE",  // todo this collides with above
        "SF3" -> "SF3",
        "SFxT" -> "SFxT",
        "GGXrdRev" -> "GGXrd Rev|GGXrd REVELATOR",
        "GGXrd" -> "GGXrd",
        "TTT2" -> "Tekken Tag Tournament 2"
    )
    val rRounds = "(?:X[0-9] )?"
    val rResolution = " *(?:1080p|720p)"

    object GameLastParser extends VideoParser {
        val regexStr = (
            rRounds +
            rPlayer +
            rCharacter +
            rVersus +
            rPlayer +
            rCharacter +
            rGame +
            rResolution
        )
        def parseSuccess(videoItem: VideoItem, regexMatch: List[String]): VideoData = {
            new VideoData(
                videoItem.id,
                videoItem.timestamp,
                fixGame(regexMatch(4)),
                fixPlayers(regexMatch(0), regexMatch(2)),
                fixCharacters(regexMatch(1), regexMatch(3))
            )
        }
    }
    object GameFirstParser extends VideoParser {
        val regexStr = (
            rGame + " *- *" +
            rRounds +
            rPlayer +
            rCharacter +
            rVersus +
            rPlayer +
            rCharacter +
            rResolution
        )
        def parseSuccess(videoItem: VideoItem, regexMatch: List[String]): VideoData = {
            new VideoData(
                videoItem.id,
                videoItem.timestamp,
                fixGame(regexMatch(0)),
                fixPlayers(regexMatch(1), regexMatch(3)),
                fixCharacters(regexMatch(2), regexMatch(4))
            )
        }
    }
    val parsers = List(GameLastParser, GameFirstParser)
}

object OlympicGamingParser extends ChannelParser {
    val channel = YouTubeChannel.OlympicGaming

    val rGameMap = ListMap(
        "SF5" -> "Street Fighter (?:5 *\\/? *V|5|V)|SFV",
        "SFxT" -> "Street Fighter X Tekken",
        "P4AU" -> "Persona 4 Arena Ultimax",
        "MKX" -> "Mortal Kombat X",
        "SC5" -> "SoulCalibur 5\\/V",
        "DOA5" -> "Dead Or Alive 5 Last Round",
        "Smash4" -> "Super Smash Bros *Wii U",
        "GGXrd" -> "Guilty Gear Xrd|Guilty Gear",
        "KI" -> "Killer Instinct"
    )
    val rEndtag = "(?:Wii|Xbox|PS4|-? ?Gameplay).*"

    object GameLastParser extends VideoParser {
        val regexStr = (
            rPlayer +
            rCharacter +
            rVersus +
            rPlayer +
            rCharacter +
            rGame +
            rEndtag
        )
        def parseSuccess(videoItem: VideoItem, regexMatch: List[String]): VideoData = {
            new VideoData(
                videoItem.id,
                videoItem.timestamp,
                fixGame(regexMatch(4)),
                fixPlayers(regexMatch(0), regexMatch(2)),
                fixCharacters(regexMatch(1), regexMatch(3))
            )
        }
    }
    object GameFirstParser extends VideoParser {
        val regexStr = (
            rGame + ": " +
            rPlayer +
            rCharacter +
            rVersus +
            rPlayer +
            rCharacter +
            rEndtag
        )
        def parseSuccess(videoItem: VideoItem, regexMatch: List[String]): VideoData = {
            new VideoData(
                videoItem.id,
                videoItem.timestamp,
                fixGame(regexMatch(0)),
                fixPlayers(regexMatch(1), regexMatch(3)),
                fixCharacters(regexMatch(2), regexMatch(4))
            )
        }
    }
    val parsers = List(GameLastParser, GameFirstParser)
}
